stream/src/main/scala/org/apache/pekko/stream/impl/fusing/CollectWhile.scala

```diff
@@ -43,18 +44,17 @@ private[pekko] final class CollectWhile[In, Out](pf: PartialFunction[In, Out]) e
       private lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
       import Collect.NotApplied
 
+      @nowarn("msg=Any")
       override final def onPush(): Unit =
         try {
           // 1. `applyOrElse` is faster than (`pf.isDefinedAt` and then `pf.apply`)
           // 2. using reference comparing here instead of pattern matching can generate less and quicker bytecode,
           //   eg: just a simple `IF_ACMPNE`, and you can find the same trick in `Collect` operator.
           //   If you interest, you can check the associated PR for this change and the
           //   current implementation of `scala.collection.IterableOnceOps.collectFirst`.
-          val result = pf.applyOrElse(grab(in), NotApplied)
-          if (result.asInstanceOf[AnyRef] eq NotApplied) {
-            completeStage()
-          } else {
-            push(out, result.asInstanceOf[Out])
+          pf.applyOrElse(grab(in), NotApplied) match {
+            case _: NotApplied.type   => completeStage()
```

Comment:
Why not match on the singleton instance instead? In any case, it seems uncommon to match on the singleton type.

It's not even easy to say if matching on the type or the reference is supposed to be faster in general, and maybe the Scala compiler should make the right choice. (On one hand, reference comparison is as fast as it gets, on the other hand, getting the singleton instance reference is at least a static field lookup which might not be faster than comparing types, would depend on the JIT compiler).

Reply:
> Why not match on the singleton instance instead? In any case, it seems uncommon to match on the singleton type.

I am under the impression it only works if you match on the type and not on the instance, @som-snytt can you confirm?

By only works I mean producing the optimal bytecode which was the point of the change in the first place.

Reply:
```scala
  def collect[A](a: A, pf: PartialFunction[A, A]): A = {
    pf.applyOrElse(a, NotApplied) match {
      case NotApplied => a
      case _          => a
    }
  }

  def collect2[A](a: A, pf: PartialFunction[A, A]): A = {
    pf.applyOrElse(a, NotApplied) match {
      case _: NotApplied.type => a
      case _                  => a
    }
  }
```

ByteCode:
![image](https://github.com/apache/incubator-pekko/assets/501740/6bd24973-6320-4947-9530-e77d18f7d105)

Reply:
@mdedetrich @jrudolph I think just as @som-snytt said , the current version should generate less bytecode.

Reply:
The clunky pattern is just to avoid a clunky `eq`.

You might use a local val to avoid "cost of obtaining the value".

This is like pattern match on List with Nil case, where match on `::` first, then just `case nil`. It's nice if you have a type to check.

An example where just "obtaining the value" is expensive is `NoSymbol` in the compiler, because of the "cake". There are edits trying to choose between `eq` and `==` but that is swamped by the fetch. It is outweighed by style concerns.

If you can measure a difference reliably, then of course you are equipped to choose.